common/src/main/java/org/opensearch/ml/common/model/MLGuard.java

```diff
@@ -0,0 +1,162 @@
+        List<StopWords> stopWords = guardrail.getStopWords();
+        if (stopWords == null || stopWords.isEmpty()) {
+            return;
+        }
+        for (StopWords e : stopWords) {
+            map.put(e.getIndex(), Arrays.asList(e.getSourceFields()));
+        }
+    }
+
+    public Boolean validate(String input, int type) {
+        switch (type) {
+            case 0: // validate input
+                return validateRegexList(input, inputRegexPattern) && validateStopWords(input, stopWordsIndicesInput);
+            case 1: // validate output
+                return validateRegexList(input, outputRegexPattern) && validateStopWords(input, stopWordsIndicesOutput);
+            default:
+                throw new IllegalArgumentException("Unsupported type to validate for guardrails.");
+        }
+    }
+
+    public Boolean validateRegexList(String input, List<Pattern> regexPatterns) {
+        for (Pattern pattern : regexPatterns) {
+            if (!validateRegex(input, pattern)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public Boolean validateRegex(String input, Pattern pattern) {
+        Matcher matcher = pattern.matcher(input);
+        return !matcher.matches();
+    }
+
+    public Boolean validateStopWords(String input, Map<String, List<String>> stopWordsIndices) {
+        for (Map.Entry entry : stopWordsIndices.entrySet()) {
+            if (!validateStopWordsSingleIndex(input, (String) entry.getKey(), (List<String>) entry.getValue())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public Boolean validateStopWordsSingleIndex(String input, String indexName, List<String> fieldNames) {
+        SearchRequest searchRequest;
+        AtomicBoolean hitStopWords = new AtomicBoolean(false);
+        String queryBody;
+        Map<String, String> documentMap = new HashMap<>();
+        for (String field : fieldNames) {
+            documentMap.put(field, input);
+        }
+        Map<String, Object> queryBodyMap = Map
+                .of("query", Map.of("percolate", Map.of("field", "query", "document", documentMap)));
+        CountDownLatch latch = new CountDownLatch(1);
+
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
+            queryBody = AccessController.doPrivileged((PrivilegedExceptionAction<String>) () -> gson.toJson(queryBodyMap));
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+            XContentParser queryParser = XContentType.JSON.xContent().createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, queryBody);
+            searchSourceBuilder.parseXContent(queryParser);
+            searchSourceBuilder.size(1); //Only need 1 doc returned, if hit.
+            searchRequest = new SearchRequest().source(searchSourceBuilder).indices(indexName);
+            context.restore();
```

Comment:
Why restore context here? Should restore in action listener

Reply:
https://github.com/opensearch-project/ml-commons/pull/2283